packages/workspace/src/tasks-runner/hasher.ts

```diff
@@ -0,0 +1,87 @@
+import { ProjectGraph, ProjectGraphNode } from '../core/project-graph';
+import { NxJson } from '../core/shared-interfaces';
+import { Task } from './tasks-runner';
+import { createHash } from 'crypto';
+import { readFileSync, statSync } from 'fs';
+
+export class Hasher {
+  implicitDependenciesHash: string;
+  hashes: { [k: string]: string } = {};
+
+  constructor(
+    private readonly projectGraph: ProjectGraph,
+    private readonly nxJson: NxJson
+  ) {
+    this.implicitDependenciesHash = this.implicitDepsHash();
+  }
+
+  hash(task: Task): string {
+    const hash = createHash('sha256');
+    hash.update(task.target.project || '');
+    hash.update(task.target.target || '');
+    hash.update(task.target.configuration || '');
+    hash.update(JSON.stringify(task.overrides));
+    hash.update(this.implicitDependenciesHash);
+    this.traverseInDepthFirstOrder(task).forEach(p => {
+      hash.update(this.hashProjectNode(p));
+    });
+    return hash.digest('hex');
+  }
+
+  private traverseInDepthFirstOrder(task: Task): ProjectGraphNode[] {
+    const r = this.traverseNode(task.target.project, []);
+    return r.map(rr => this.projectGraph.nodes[rr]);
+  }
+
+  private traverseNode(project: string, acc: string[]): string[] {
+    if (acc.indexOf(project) > -1) return acc;
+    acc.push(project);
+    (this.projectGraph.dependencies[project] || [])
+      .map(t => t.target)
+      .forEach(r => {
+        this.traverseNode(r, acc);
+      });
+    return acc;
+  }
+
+  private hashProjectNode(p: ProjectGraphNode): string {
+    if (this.hashes[p.name]) {
+      return this.hashes[p.name];
+    } else {
+      const hash = createHash('sha256');
+      p.data.files.forEach(f => {
+        hash.update(this.readFileContents(f.file));
+      });
+      const r = hash.digest('hex');
+      this.hashes[p.name] = r;
+      return r;
+    }
+  }
+
+  private readFileContents(path: string) {
+    const stats = statSync(path);
+    const fileSizeInMegabytes = stats.size / 1000000.0;
```

Comment:
Does the `.0` do something differently? Just curious since it does in other languages.

Reply:
Doesn't. Removed it.